src/natcap/invest/urban_nature_access.py

```diff
@@ -945,95 +1467,206 @@ def _filter_population(population, greenspace_budget, numpy_filter_op):
+    source_vector = ogr.Open(source_aoi_vector_path)
     driver = ogr.GetDriverByName('GPKG')
     driver.CopyDataSource(source_vector, target_aoi_vector_path)
     source_vector = None
 
     target_vector = gdal.OpenEx(target_aoi_vector_path, gdal.GA_Update)
     target_layer = target_vector.GetLayer()
 
-    supply_sum_fieldname = 'SUP_DEMadm_cap'
-    undersupply_fieldname = 'Pund_adm'
-    oversupply_fieldname = 'Povr_adm'
-    target_fieldnames = []
-    population_groupnames = []
-    for defn in target_layer.schema:
-        name = defn.GetName()
-        if not re.match(POP_FIELD_REGEX, name):
-            continue
-        groupname = re.sub(POP_FIELD_REGEX, '', name)
-        population_groupnames.append(groupname)
-        target_fieldnames.append(f'{oversupply_fieldname}_{groupname}')
-        target_fieldnames.append(f'{undersupply_fieldname}_{groupname}')
-
-    for fieldname in (supply_sum_fieldname,
-                      undersupply_fieldname,
-                      oversupply_fieldname,
-                      *target_fieldnames):
+    for fieldname in next(iter(feature_attrs.values())).keys():
         field = ogr.FieldDefn(fieldname, ogr.OFTReal)
         field.SetWidth(24)
         field.SetPrecision(11)
         target_layer.CreateField(field)
 
-    greenspace_stats = pygeoprocessing.zonal_statistics(
-        (greenspace_budget_path, 1), target_aoi_vector_path)
-    population_stats = pygeoprocessing.zonal_statistics(
-        (population_path, 1), target_aoi_vector_path)
-    undersupplied_stats = pygeoprocessing.zonal_statistics(
-        (undersupplied_populations_path, 1), target_aoi_vector_path)
-    oversupplied_stats = pygeoprocessing.zonal_statistics(
-        (oversupplied_populations_path, 1), target_aoi_vector_path)
-
     target_layer.StartTransaction()
     for feature in target_layer:
         feature_id = feature.GetFID()
+        for attr_name, attr_value in feature_attrs[feature_id].items():
+            feature.SetField(attr_name, attr_value)
 
-        avg_greenspace_supply_demand = (
-            greenspace_stats[feature_id]['sum'] /
-            population_stats[feature_id]['sum'])
-        feature.SetField(supply_sum_fieldname, avg_greenspace_supply_demand)
-        undersupply = undersupplied_stats[feature_id]['sum']
-        oversupply = oversupplied_stats[feature_id]['sum']
-        feature.SetField(undersupply_fieldname, undersupply)
-        feature.SetField(oversupply_fieldname, oversupply)
-        for pop_groupname in population_groupnames:
-            group_proportion = feature.GetField(f'pop_{pop_groupname}')
-            feature.SetField(
-                f'{oversupply_fieldname}_{pop_groupname}',
-                oversupply * group_proportion)
-            feature.SetField(
-                f'{undersupply_fieldname}_{pop_groupname}',
-                undersupply * group_proportion)
         target_layer.SetFeature(feature)
     target_layer.CommitTransaction()
```

Comment:
I feel like we usually do a little clean up after having a vector or raster open. Should we set the target vector and layer to None?

Reply:
Might as well! 868213d6d